src/handlers/IntentHandler.sol

```diff
@@ -0,0 +1,250 @@
+    if (inputs.accountAndSubAccountIds.length != inputs.cmds.length) revert IntentHandler_BadLength();
+
+    ExecuteIntentVars memory _localVars;
+    Command _cmd;
+    ExecuteTradeOrderVars memory _vars;
+    bytes32 key;
+
+    // Update price to Pyth
+    pyth.updatePriceFeeds(inputs.priceData, inputs.publishTimeData, inputs.minPublishTime, inputs.encodedVaas);
+
+    _localVars.cmdsLength = inputs.accountAndSubAccountIds.length;
+    _localVars.tpTokens = configStorage.getHlpTokens();
+
+    for (uint256 _i; _i < _localVars.cmdsLength; ) {
+      _localVars.mainAccount = address(uint160(inputs.accountAndSubAccountIds[_i].decodeUint(0, 160)));
+      _localVars.subAccountId = uint8(inputs.accountAndSubAccountIds[_i].decodeUint(160, 8));
+      _cmd = Command(inputs.cmds[_i].decodeUint(0, 3));
+
+      if (_cmd == Command.ExecuteTradeOrder) {
+        _vars.order.marketIndex = inputs.cmds[_i].decodeUint(3, 8);
+        _vars.order.sizeDelta = inputs.cmds[_i].decodeInt(11, 54) * 1e22;
+        _vars.order.triggerPrice = inputs.cmds[_i].decodeUint(65, 54) * 1e22;
+        _vars.order.acceptablePrice = inputs.cmds[_i].decodeUint(119, 54) * 1e22;
+        _vars.order.triggerAboveThreshold = inputs.cmds[_i].decodeBool(173);
+        _vars.order.reduceOnly = inputs.cmds[_i].decodeBool(174);
+        _vars.order.tpToken = _localVars.tpTokens[uint256(inputs.cmds[_i].decodeUint(175, 7))];
+        _vars.order.createdTimestamp = inputs.cmds[_i].decodeUint(182, 32);
+        _vars.order.expiryTimestamp = inputs.cmds[_i].decodeUint(214, 32);
+        _vars.order.account = _localVars.mainAccount;
+        _vars.order.subAccountId = _localVars.subAccountId;
+
+        key = keccak256(abi.encode(inputs.accountAndSubAccountIds[_i], inputs.cmds[_i]));
+        if (executedIntents[key]) {
+          revert IntentHandler_IntentReplay();
+        }
+
+        _validateSignature(_vars.order, inputs.signatures[_i], _localVars.mainAccount);
+        _collectExecutionFeeFromCollateral(_localVars.mainAccount, _localVars.subAccountId);
+        bool _isSuccess = _executeTradeOrder(_vars);
+
+        // If the trade order is executed successfully, record the order as executed
+        if (_isSuccess) executedIntents[key] = true;
+      }
+
+      unchecked {
+        ++_i;
+      }
+    }
+  }
+
+  function _executeTradeOrder(ExecuteTradeOrderVars memory vars) internal returns (bool isSuccess) {
+    // try executing order
+    try tradeOrderHelper.execute(vars) {
+      // Execution succeeded
+      return true;
+    } catch Error(string memory errMsg) {
+      _handleOrderFail(vars, bytes(errMsg));
+    } catch Panic(uint /*errorCode*/) {
+      _handleOrderFail(vars, bytes("Panic occurred while executing trade order"));
+    } catch (bytes memory errMsg) {
+      _handleOrderFail(vars, errMsg);
+    }
+    return false;
+  }
+
+  function _collectExecutionFeeFromCollateral(address _primaryAccount, uint8 _subAccountId) internal {
```

Comment:
Better to put this in the `Service` level instead of allow Handler to alter the Storage directly. Might introduce a new service? `GasService`?

Reply:
done